core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -1694,16 +1696,18 @@ public static String toString(double x) {
     BigDecimal bigDecimal =
         new BigDecimal(x, MathContext.DECIMAL64).stripTrailingZeros();
     final String s = bigDecimal.toString();
-    return s.replaceAll("0*E", "E").replace("E+", "E");
+    return PATTERN_0_STAR_E.matcher(s).replaceAll("E").replace("E+", "E");
   }
 
   /** CAST(DECIMAL AS VARCHAR). */
   public static String toString(BigDecimal x) {
     final String s = x.toString();
-    if (s.startsWith("0")) {
+    if (s.equals("0")) {
+      return s;
+    } else if (s.startsWith("0.")) {
```

Comment:
Is this locale-dependent?
The use of dot here does not look right.

Reply:
That may or may not be a bug. If there's a bug (e.g. you can reproduce in SQL) please log it. I refactored the existing code, so I don't think I made anything worse.

Reply:
The `Float.toString()`is not locale-dependent - that is, it always uses '.' as decimal separator - and I think that's what we want for `CAST(REAL AS VARCHAR)`.